File: 3. Behavioral Design Patterns/Mediator/Javascript/MediatorExample.js
// Mediator Pattern Example in JavaScript
// Chat room where users send messages through a mediator

class ChatRoom {
  sendMessage(message, user) {}
  addUser(user) {}
}

class User {
  constructor(name, chatRoom) { this.name = name; this.chatRoom = chatRoom }
  send(message) {}
  receive(message, from) {}
}

class ConcreteUser extends User {
  send(message) { this.chatRoom.sendMessage(message, this) }
  receive(message, from) { console.log(`${this.name} received: ${message} (from ${from})`) }
}

class ConcreteChatRoom extends ChatRoom {
  constructor() { super(); this.users = [] }
  addUser(user) { this.users.push(user) }
  sendMessage(message, sender) {
    for (const user of this.users)
      if (user !== sender)
        user.receive(message, sender.name)
  }
}

;(function main() {
  const chatRoom = new ConcreteChatRoom()
  const alice = new ConcreteUser('Alice', chatRoom)
  const bob = new ConcreteUser('Bob', chatRoom)
  const charlie = new ConcreteUser('Charlie', chatRoom)
  chatRoom.addUser(alice)
  chatRoom.addUser(bob)
  chatRoom.addUser(charlie)

  alice.send('Hello everyone!')
  bob.send('Hi Alice!')
})()

File: 3. Behavioral Design Patterns/Mediator/Java/MediatorExample.java
// Mediator Pattern Example in Java
// Chat room where users send messages through a mediator

import java.util.*;

interface ChatRoom { void sendMessage(String message, User user); void addUser(User user); }

abstract class User {
    protected ChatRoom chatRoom;
    protected String name;
    public User(String name, ChatRoom room) { this.name = name; this.chatRoom = room; }
    public abstract void send(String message);
    public abstract void receive(String message, String from);
    public String getName() { return name; }
}

class ConcreteUser extends User {
    public ConcreteUser(String name, ChatRoom room) { super(name, room); }
    public void send(String message) { chatRoom.sendMessage(message, this); }
    public void receive(String message, String from) { System.out.println(name + " received: " + message + " (from " + from + ")"); }
}

class ConcreteChatRoom implements ChatRoom {
    private final List<User> users = new ArrayList<>();
    public void addUser(User user) { users.add(user); }
    public void sendMessage(String message, User sender) {
        for (User user : users)
            if (user != sender)
                user.receive(message, sender.getName());
    }
}

public class MediatorExample {
    public static void main(String[] args) {
        ChatRoom chatRoom = new ConcreteChatRoom();
        User alice = new ConcreteUser("Alice", chatRoom);
        User bob = new ConcreteUser("Bob", chatRoom);
        User charlie = new ConcreteUser("Charlie", chatRoom);
        chatRoom.addUser(alice);
        chatRoom.addUser(bob);
        chatRoom.addUser(charlie);

        alice.send("Hello everyone!");
        bob.send("Hi Alice!");
    }
}
